### PR TITLE
Allow Netherite provider to be implemented separately

### PR DESCRIPTION
**Pull request by sebastianburckhardt**
Support for using the EventSourced backend, as in the work-in-progress PR azure/durabletask#461.

Besides adding a durability provider, the changes include:
- turn off message sorter logic for backend that delivers messages in order
- support poll-free version of WaitForOrchestration for backends that support client responses
- change tracing options for unit tests

### CODE REVIEW

**Comment by davidmrdavid** [comment]
@sebastianburckhardt , if this is WIP, should we consider making it a draft PR? :)

**Comment by sebastianburckhardt** [comment]
Not sure what the difference is between draft and WIP?

Since we just decided to restructure the projects I can't be quite sure what will change. Currently figuring this out. My guess is that in its eventual form, this PR will stay similar, as far as modifications to existing files are concerned, but it will not include the new durability provider.

**Comment by davidmrdavid** [comment]
@sebastianburckhardt , a _draft_ is a GitHub UI feature to acknowledge that a PR is work-in-progress. I've used it in the past to make a clean separation between PRs that need immediate reviews for immediate merge consideration and the PRs that do not. You can read about it here: 
https://github.blog/2019-02-14-introducing-draft-pull-requests/

Essentially if you CTRL+F "draft" in this page, you should be able to convert this into a draft PR. Just a suggestion!

**Comment by sebastianburckhardt** [comment]
This is now ready to be reviewed. It only contains the changes needed for implementing Netherite as an external NuGet package. The changes include:

- turn off message sorter logic for backend that delivers messages in order
- support poll-free version of WaitForOrchestration for backends that support client responses
- add three protected methods to DurabilityProvider that allow implementations to access functionality that is otherwise internal

File: src/WebJobs.Extensions.DurableTask/ContextImplementations/DurableClient.cs
Comment (davidmrdavid) [inline src/WebJobs.Extensions.DurableTask/ContextImplementations/DurableClient.cs:74]: I realize this might be trivial but, since this is a public method, shouldn't it at least inherit its documentation?

File: src/WebJobs.Extensions.DurableTask/ContextImplementations/DurableClient.cs
Comment (davidmrdavid) [inline src/WebJobs.Extensions.DurableTask/ContextImplementations/DurableClient.cs:538]: nitpick: since this is such a long conditional check, I would recommend splitting it into smaller, well-named, boolean variables. It would improve readability and help us detect bugs in the future

File: src/WebJobs.Extensions.DurableTask/HttpApiHandler.cs
Comment (davidmrdavid) [inline src/WebJobs.Extensions.DurableTask/HttpApiHandler.cs:183]: Why don't we take an action in case of a Timeout? It would be good to document this :)

File: src/WebJobs.Extensions.DurableTask/HttpApiHandler.cs
Comment (davidmrdavid) [inline src/WebJobs.Extensions.DurableTask/HttpApiHandler.cs:189]: Can we document / comment about the difference between these two execution paths?

File: src/WebJobs.Extensions.DurableTask/HttpApiHandler.cs
Comment (davidmrdavid) [inline src/WebJobs.Extensions.DurableTask/HttpApiHandler.cs:174]: The syntax of `client is DurableClient durableClient` confuses me a bit. I would have expected this to be simply `client is DurableClient`, which would be a type-check. What is this achieving?

File: src/WebJobs.Extensions.DurableTask/HttpApiHandler.cs
Comment (davidmrdavid) [inline src/WebJobs.Extensions.DurableTask/HttpApiHandler.cs:726]: Can we document how this else-if is different from the pre-existing if-statement above? They seem quite similar. 

In fact, there might be a refactor we we only check for `if (timeout.HasValue)` and inside of that we switch the final parameter to `WaitForCompletionOrCreateCheckStatusResponseAsync` depending on some differentiating condition.

I do not feel strongly about a refactor like that, but I would recommend a quick comment explaining the difference. Thanks!

**Review by davidmrdavid — COMMENTED** [review]
Looks pretty good to me, I give it my thumbs-up but I do have suggestions for improvements, please see my comments.

I'll wait for a second pair of eyes to give this the approval 😄

**Comment by davidmrdavid** [comment]
Ah it also seems like there's a merge conflict with `src/WebJobs.Extensions.DurableTask/DurableTaskExtension.cs`

File: src/WebJobs.Extensions.DurableTask/ContextImplementations/DurableClient.cs
Comment (sebastianburckhardt) [reply src/WebJobs.Extensions.DurableTask/ContextImplementations/DurableClient.cs:74]: yes, it should. Will fix.

File: src/WebJobs.Extensions.DurableTask/HttpApiHandler.cs
Comment (sebastianburckhardt) [reply src/WebJobs.Extensions.DurableTask/HttpApiHandler.cs:174]: the name 'durableClient' gets bound and is used right after the &&.

File: src/WebJobs.Extensions.DurableTask/ContextImplementations/DurableClient.cs
Comment (sebastianburckhardt) [reply src/WebJobs.Extensions.DurableTask/ContextImplementations/DurableClient.cs:538]: Yes, I agree. I am splitting this apart now.

File: src/WebJobs.Extensions.DurableTask/HttpApiHandler.cs
Comment (sebastianburckhardt) [reply src/WebJobs.Extensions.DurableTask/HttpApiHandler.cs:183]: Yes, adding comment.

File: src/WebJobs.Extensions.DurableTask/HttpApiHandler.cs
Comment (sebastianburckhardt) [reply src/WebJobs.Extensions.DurableTask/HttpApiHandler.cs:189]: Yes.

File: src/WebJobs.Extensions.DurableTask/HttpApiHandler.cs
Comment (sebastianburckhardt) [reply src/WebJobs.Extensions.DurableTask/HttpApiHandler.cs:726]: I changed the logic for this a bit, and added a comment.

**Review by davidmrdavid — COMMENTED** [review]
Really happy with the latest changes. LGTM! Still waiting on a 2nd pair of eyes just in case

**Comment by davidmrdavid** [comment]
Re-started the CI build checks since it appears to have been running for over an hour. Sometimes this happens, _for some reason,_ with the CI. Hoping they pass now!

File: src/WebJobs.Extensions.DurableTask/DurabilityProvider.cs
Comment (ConnorMcMahon) [inline src/WebJobs.Extensions.DurableTask/DurabilityProvider.cs:444]: Is there a reason this is now a `protected` method, as opposed to having people just directly call `EntityId.GetScheduleIdFromEntityId`? If it's because we want to allow each provider to specify their own mapping, then shouldn't this be `virtual`?

File: src/WebJobs.Extensions.DurableTask/DurableTaskExtension.cs
Comment (ConnorMcMahon) [inline src/WebJobs.Extensions.DurableTask/DurableTaskExtension.cs:1]: Would it be possible to undo the whitespace diffs here? It makes it difficult to remove the changes in this file.

File: src/WebJobs.Extensions.DurableTask/DurabilityProvider.cs
Comment (ConnorMcMahon) [inline src/WebJobs.Extensions.DurableTask/DurabilityProvider.cs:480]: Similarly here, I'm not sure I love the idea of adding protected methods to give us access to some internal static method.  Would it be easier to just to make these static methods public (and maybe put them in a more aptly named class if other storage providers need access to this method)?

**Review by ConnorMcMahon — COMMENTED** [review]
Overall looks fairly good, just a few questions.

File: src/WebJobs.Extensions.DurableTask/DurabilityProvider.cs
Comment (sebastianburckhardt) [reply src/WebJobs.Extensions.DurableTask/DurabilityProvider.cs:444]: It is a private method, so the point was to make it "semi-public", meaning accessible to implementors of DurabilityProvider. That way it does not pollute the public API, e.g. normal users won't see it in IntelliSense.

File: src/WebJobs.Extensions.DurableTask/DurabilityProvider.cs
Comment (sebastianburckhardt) [reply src/WebJobs.Extensions.DurableTask/DurabilityProvider.cs:480]: I suppose I did not think of a case where anybody other than the DurabilityProvider implementor needs access.  Though I guess AzureStorage provider is an example where one may want access without going via a DurabilityProvider. Perhaps a static class would be better, such as ProviderUtils so users know not to look at it unless they implement a provider. I will do that.

File: src/WebJobs.Extensions.DurableTask/DurableTaskExtension.cs
Comment (sebastianburckhardt) [reply src/WebJobs.Extensions.DurableTask/DurableTaskExtension.cs:1]: Not sure what is happening here. I don't see this problem on my local checkout - which shows a normal diff.

Comment (ConnorMcMahon) [reply src/WebJobs.Extensions.DurableTask/DurableTaskExtension.cs:1]: If that doesn't work, maybe the quickest workaround would just be to add a brief GitHub PR comment next to where the largest changes are so we know where to look.

Comment (sebastianburckhardt) [reply src/WebJobs.Extensions.DurableTask/DurableTaskExtension.cs]: I think it worked, the diff looks normal now.

File: src/WebJobs.Extensions.DurableTask/DurabilityProvider.cs
Comment (sebastianburckhardt) [reply src/WebJobs.Extensions.DurableTask/DurabilityProvider.cs]: I have pushed this change also.

**Review by ConnorMcMahon — APPROVED** [review]
LGTM!

**Pull request merged**
